targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/api/trng_api.c

```diff
@@ -52,44 +52,58 @@
 #define TRNG_CNT_VAL	4095
 #define TRNG_PRESCALER 	2
 
-static ADI_RNG_HANDLE     RNGhDevice;                     /* Memory to handle CRC Device */
-static uint32_t           RngDevMem[(ADI_RNG_MEMORY_SIZE + 3)/4]; /* Data buffers for Random numbers */
+/* Data buffers for Random numbers */
+static uint32_t RngDevMem[(ADI_RNG_MEMORY_SIZE + 3)/4];
 
 void trng_init(trng_t *obj)
 {
-    (void)obj;
+	ADI_RNG_HANDLE RNGhDevice;
+
+	// Open the device
     adi_rng_Open(0,RngDevMem,sizeof(RngDevMem),&RNGhDevice);
 
     // Set sample length for the H/W RN accumulator
     adi_rng_SetSampleLen(RNGhDevice, TRNG_PRESCALER, TRNG_CNT_VAL);
 
-    // Enable the RNG
+	// Disable buffering - single byte generation only
+	adi_rng_EnableBuffering(RNGhDevice, false);
+
+	// Enable the TRNG
     adi_rng_Enable(RNGhDevice, true);
+
+	// Save device handle
+	obj->RNGhDevice = RNGhDevice;
 }
 
 void trng_free(trng_t *obj)
 {
-    (void)obj;
+    ADI_RNG_HANDLE RNGhDevice = obj->RNGhDevice;
+
     adi_rng_Enable(RNGhDevice, false);
     adi_rng_Close(RNGhDevice);
 }
 
 int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
 {
-    (void)obj;
+    ADI_RNG_HANDLE RNGhDevice = obj->RNGhDevice;
     bool bRNGRdy;
     uint32_t nRandomNum, i;
 
-    for (i = 0; i < length; ) {
-        // wait for the RNG ready to give a random number
-        adi_rng_GetRdyStatus(RNGhDevice, &bRNGRdy);
-        if (bRNGRdy) {
-            // read the random number
-            adi_rng_GetRngData(RNGhDevice, &nRandomNum);
-            output[i++] = (uint8_t) nRandomNum;
-        }
+    for (i = 0; i < length; i++) {
+        // Loop until the device has data to be read
+        do {
+            adi_rng_GetRdyStatus(RNGhDevice, &bRNGRdy);
+        } while (!bRNGRdy);
+
+        // Read the RNG
+        adi_rng_GetRngData(RNGhDevice, &nRandomNum);
+
+        // Save the output
+		output[i] = (uint8_t)nRandomNum;
```

Comment:
same comment as before. `nRandomNum` is `uint32_t`, and you are taking only 8 bits from it. THis doesn't gurantee sufficient entropy. I suggest you do the loop until `i < length/sizeof(uint32_t)`, and the `memcpy(((uint32_t*)output)[i],&nRandomNum, sizeof(uint32_t))` , and after the loop copy the residue (length%sizeof(uint32_t) bytes )
This will also result in fewer calls to Hardware, possibly improving performance

Reply:
what is the return type of the `adi_xxx` functions? Can they fail?

Reply:
Sorry I failed to mention. The function adi_rng_GetRngData outputs either an 8-bit byte or a 32-bit word, depending on whether the TRNG buffering mode is enabled. By default the buffering mode is disabled and it would output an 8-bit byte, so only the LSB in nRandomNum is valid. In the trng_init, I added code to make sure the buffering mode is disabled. 

The return values of the adi_xxx functions indicate whether the operation is success or had issues. In the case of adi_rng_GetRngData and adi_rng_GetRdyStatus calls, they would only return an error if the TRNG is not initialized in which case the device handle is invalid.

Reply:
>By default the buffering mode is disabled and it would output an 8-bit byte, so only the LSB in nRandomNum is valid.
I would use masking instead of casting
>The return values of the adi_xxx functions indicate whether the operation is success or had issues.
In this case, it is better to check return type, and fail if the internal `adi_xxx` fail

Reply:
Will add error checking for the trng_get_byte function.